MTV-2919: Allow users to change the non-conformant VM names on the fly

avivtur opened the pull request:
## 📝 Links

https://issues.redhat.com/browse/MTV-2919

## 📝 Description

Adding an action to add a targetName for VMs in a plan

## 🎥 Demo


https://github.com/user-attachments/assets/2059b420-f6f5-475b-b0c5-f5ade95f1366



## 📝 CC://

@mmenestr @heyethankim please review

avivtur commented:
> One more thing, for consistency, I just got feedback from UX content writers that we should have a question mark at the end of these Edit modal headers, and that we should specify what we are saving in each modal in the button. Are you able to make that change too? This is what the modal would look like in the end, with updates to the header and the button:
> 
> <img alt="image" width="539" height="386" src="https://private-user-images.githubusercontent.com/51165119/473562470-84d4372d-5421-45fa-8676-1b5fa05ae229.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQzMDI5MjcsIm5iZiI6MTc1NDMwMjYyNywicGF0aCI6Ii81MTE2NTExOS80NzM1NjI0NzAtODRkNDM3MmQtNTQyMS00NWZhLTg2NzYtMWI1ZmEwNWFlMjI5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODA0VDEwMTcwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJjNmVmYzc3OTE3OTMxOWUzMTViYTJlNGMzNmU2Yjk4NDMyZmFkZjA3MTVjODIwODBiNTdlMzYxNTM1MGJjZWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.2UIYrrCyFT_OmkLaW6Qd29Rk92Xz4m-iInRiAxOEW1k">

@mmenestr Hi :) 
Currently all modals don't have a question mark and the save button just states "save", if we want to change that across all modals, than it should be in a separated epic. And doing just this modal in this PR will make it inconsistent with the rest.

 

> This looks good! I just have a couple comments from a consistency standpoint: I would have the words in the menu item match the modal header.
> 
> * Instead of "Set target name" as the modal header, I'd have it say "Edit target name" 

good catch +1

> and have the text input be pre-filled with the current name of the VM (assuming that is what it would be called after migration if the user did not set this). From there, users can make edits or leave it as is.

This part is a bit trickey, because some names aren't fitting the kubernetes name pattern allowed, and the BE is doing some manipulation to it if the target name is unset. trying to mimic the BE mechanism would most defentily cause bugs in the future, so I'd avoid with pre-filled input if there's none.

> * I would remove the word "VM" from the menu item and just say "Edit target name"

+1 

> * I would also update the description of the modal to say "Enter the name you'd like the **esxi8-vm** VM to have after migration." just to make it more concise !

+1

jpuzz0 reviewed (commented):
Looks good overall. Added a few comments.

jschuler commented:
General question, all those things that are editable, wouldn't it make sense to show columns for them (maybe optionally by clicking Manage columns)? Seems like being able to edit something without showing the edited value in UI is not good.
Otherwise a user may as well edit it directly in YAML only IMO :) @mmenestr

jschuler commented on src/plans/details/tabs/VirtualMachines/components/PlanSpecVirtualMachinesList/components/VirtualMachineTargetName/EditVirtualMachineTargetName.tsx:
Saw similar issue addressed elsewhere before too
https://github.com/kubev2v/forklift-console-plugin/pull/1669

avivtur replied:
@jpuzz0 @jschuler 
Did you see it happens recently? if so we need to handle it seperately

avivtur replied:
I've added a fix for this in the PR, if you could also play with it, it would be great, great catch!

avivtur commented on src/plans/details/tabs/VirtualMachines/components/PlanSpecVirtualMachinesList/components/VirtualMachineTargetName/EditVirtualMachineTargetName.tsx:
```diff
@@ -0,0 +1,64 @@
+  const [errorMessage, setErrorMessage] = useState<string | null>(null);
+
+  return (
+    <ModalForm
```
ModalForm was created just to simplify modals and their error handling. (the previous component we had for modals was very specific and had minimal flexability when implementing a new modal)
There's no real need to convert it to use react-hook-form for now as most dialogs are very simple.
If you think we need to enhance this component, I'll be happy to hear your thought on this

avivtur replied:
Fair points, Let's do this in a separated story, and since there's no rush on this for now, we'll do it after all bigger issues are solved.
Thanks for pointing this out!

avivtur commented on src/plans/details/tabs/VirtualMachines/components/PlanSpecVirtualMachinesList/components/utils/utils.ts:
```diff
@@ -84,3 +88,40 @@ export const onConfirmVirtualMachinePVCNameTemplate =
     return result;
   };
+
+export const setVMTargetName: SetVMTargetName = async ({ newValue, resource, vmIndex }) => {
```
changed to `patchVMTargetName` , if I misunderstood your intention please clearify

avivtur commented on src/plans/details/tabs/VirtualMachines/components/PlanSpecVirtualMachinesList/components/VirtualMachineTargetName/EditVirtualMachineTargetName.tsx:
```diff
@@ -0,0 +1,64 @@
+            helperTextInvalid={errorMessage}
+            label={t('VM target name')}
+          >
+            <Grid span={6}>
```
it was to match the design, but I agree it's not needed and we should have a pattern for all inputs in modals.
removed for now

avivtur commented on src/plans/details/tabs/VirtualMachines/components/PlanSpecVirtualMachinesList/components/utils/utils.ts:
```diff
@@ -84,3 +88,40 @@ export const onConfirmVirtualMachinePVCNameTemplate =
+
+  if (!validateK8sName(value)) {
+    return t(
+      "VM target name must contain only lowercase alphanumeric characters or '-', and must start or end with lowercase alphanumeric character.",
```
this text is appearing in few places, we need to standartize these validations in a single place, not in this PR

avivtur commented:
> General question, all those things that are editable, wouldn't it make sense to show columns for them (maybe optionally by clicking Manage columns)? Seems like being able to edit something without showing the edited value in UI is not good. Otherwise a user may as well edit it directly in YAML only IMO :) @mmenestr

Good question, @mmenestr @heyethankim , thoughts on what @jschuler comment?

avivtur commented on src/plans/details/tabs/VirtualMachines/components/PlanSpecVirtualMachinesList/components/VirtualMachineTargetName/EditVirtualMachineTargetName.tsx:
```diff
@@ -0,0 +1,64 @@
+const EditVirtualMachineTargetName: FC<EditVirtualMachineTargetNameProps> = ({ plan, vmIndex }) => {
+  const { t } = useForkliftTranslation();
+  const [inputValue, setInputValue] = useState(
+    (plan?.spec?.vms as EnhancedPlanSpecVms[])?.[vmIndex]?.targetName ?? '',
```
I used to getPlanVirtualMachines for this, so the type remains the same

jpuzz0 reviewed (approved):
Ideally, we fix this issue as a part of this feature, but will not consider it blocking https://github.com/kubev2v/forklift-console-plugin/pull/1747#discussion_r2260229492

avivtur commented:
> Ideally, we fix this issue as a part of this feature, but will not consider it blocking [#1747 (comment)](https://github.com/kubev2v/forklift-console-plugin/pull/1747#discussion_r2260229492)

Will investigate

avivtur commented:
@jschuler @jpuzz0 thanks for your reviews and comments!
Added column for VM target name, and fixed the closing modal/rows on plan change.
Latest iteration demo:

https://github.com/user-attachments/assets/c2a5ee8a-d258-4a4b-847f-a52b803891bb

mmenestr commented:
I agree adding a column for each of these is a good idea, and @jschuler and I had discussed doing that, but like was said here, perhaps as part of a bigger change. But you guys can figure that part out.

This looks good but can we make the column not include the words "VM", so that it matches everything else - it should just say "Target name"

jpuzz0 commented:
Looks like this validation needs to exclude the row that is being edited. Where I wasn't expecting an error for the 1 row with the same name I just saved:

https://github.com/user-attachments/assets/f3c900db-fd9a-44ac-855f-3bc3beaeef6a

avivtur commented:
@mmenestr @jpuzz0 Thanks for reviewing again!
Fixed both of your comments

jpuzz0 commented:
LGTM
